### PR TITLE
feat: enhance file_edit tool with unified diff visualization and smart folding

### PR DESCRIPTION
**Pull request by obra**
## Summary
- Enhanced file_edit tool renderer with unified diff visualization for multi-edit operations
- Fixed corrupted diff display by replacing naive algorithm with improved look-ahead matching
- Added automatic folding for sections with 4+ consecutive unchanged lines
- Enhanced tool description with clear examples and best practices

## Test plan
- [x] Test multi-edit operations display single unified diff
- [x] Verify auto-folding works for large unchanged sections  
- [x] Confirm enhanced error display with validation details
- [x] Check tool description examples are clear and helpful
- [x] All existing tests pass with updated expectations

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

* **New Features**
  * Introduced support for multiple atomic text edits in a single file operation, with strict occurrence validation and dry run mode.
  * Enhanced error feedback with detailed validation errors, line numbers, suggestions, and file previews for unmatched or mismatched edits.
  * Improved diff viewer with automatic folding of large unchanged sections and clearer diff visualization in both unified and side-by-side views.
  * Technical details panel now displays tool call result JSON for greater transparency.

* **Bug Fixes**
  * Ensured file encoding and line endings are preserved during edits.
  * Fixed handling of empty files, non-existent files, and large files for robustness.

* **Documentation**
  * Added comprehensive implementation plan and design specification for the new file editing tool, including usage, error handling, and migration steps.

* **Tests**
  * Expanded and refactored test suites to cover multi-edit scenarios, occurrence validation, dry run mode, edge cases, and performance.
  * Added tests confirming real file modifications and detailed error reporting.

* **Refactor**
  * Unified edit API to use an array of edits, updated schemas, and improved internal diff and validation algorithms for reliability and maintainability.

* **Style**
  * Improved UI feedback for errors, successes, and dry runs with more precise messaging and visual enhancements in diff rendering.

### CODE REVIEW

File: src/tools/implementations/file-edit.ts [hunk 1]
```diff
@@ -1,18 +1,29 @@
...
+import { FilePath } from '~/tools/schemas/common';
 
-const fileEditSchema = z.object({
-  path: FilePath,
+// Define schemas for input validation
+const editOperationSchema = z.object({
   old_text: z.string(),
   new_text: z.string(),
+  occurrences: z.number().int().positive().optional(),
 });
```
[the remark this answers is not in the document]
Comment (obra) [reply src/tools/implementations/file-edit.ts]: this is wrong. non-empty strings are acceptable
